ckanext/security/authenticator.py

```diff
@@ -105,15 +105,21 @@ def authenticate(self, environ, identity):
             # Increment the throttle counter if the login failed.
             throttle.increment()
 
-        # if the CKAN authenticator has successfully authenticated
-        # the request and the user wasn't locked out above,
-        # then check the TOTP parameter to see if it is valid
-        if auth_user_name is not None:
-            totp_success = self.authenticate_totp(environ, auth_user_name)
-            # if TOTP was successful -- reset the log in throttle
-            if totp_success:
-                throttle.reset()
-                return totp_success
+        # totp authentication is enabled by default for all users
+        # totp can be disabled, if needed, by setting
+        # ckanext.security.enable_totp to false in configurations
+        if security_enable_totp():
+            # if the CKAN authenticator has successfully authenticated
+            # the request and the user wasn't locked out above,
+            # then check the TOTP parameter to see if it is valid
+            if auth_user_name is not None:
+                totp_success = self.authenticate_totp(environ, auth_user_name)
+                # if TOTP was successful -- reset the log in throttle
+                if totp_success:
+                    throttle.reset()
```

Comment:
I didn't notice this earlier, but I think you need to call throttle.reset in the case where the totp is disabled as well. Otherwise the rest will never occur when a user successfully logs in with username and password only

Reply:
good point @markstuart. I have reset throttle if security_enable_totp() returns false